kv/bolt/client.go

```diff
@@ -20,155 +21,330 @@ const (
+func (b *Bucket) NextSequence() (uint64, error) {
+	return b.bucket.NextSequence()
+}
+
+// ForwardCursor retrieves a cursor for iterating through the entries
+// in the key value store in a given direction (ascending / descending).
+func (b *Bucket) ForwardCursor(seek []byte) (kv.ForwardCursor, error) {
+	var (
+		cursor     = b.bucket.Cursor()
+		key, value = cursor.Seek(seek)
+	)
+
+	return &Cursor{
+		cursor: cursor,
+		key:    key,
+		value:  value,
+	}, nil
+}
+
+// Cursor retrieves a cursor for iterating through the entries
+// in the key value store.
+func (b *Bucket) Cursor() (kv.Cursor, error) {
+	return &Cursor{
+		cursor: b.bucket.Cursor(),
+	}, nil
+}
+
+// ForEach executes a function for each key/value pair in a bucket.
+// If the provided function returns an error then the iteration is stopped and
+// the error is returned to the caller. The provided function must not modify
+// the bucket; this will result in undefined behavior.
+func (b *Bucket) ForEach(fn func(k, v []byte) error) error {
+	return b.bucket.ForEach(fn)
+}
+
+// Cursor is a struct for iterating through the entries
+// in the key value store.
+type Cursor struct {
+	cursor *bolt.Cursor
+
+	// previously seeked key/value
+	key, value []byte
+
+	closed bool
+}
+
+// Close sets the closed to closed
+func (c *Cursor) Close() error {
+	c.closed = true
 
 	return nil
 }
 
+// Seek seeks for the first key that matches the prefix provided.
+func (c *Cursor) Seek(prefix []byte) ([]byte, []byte) {
+	if c.closed {
+		return nil, nil
+	}
+
+	k, v := c.cursor.Seek(prefix)
+	if len(k) == 0 && len(v) == 0 {
+		return nil, nil
+	}
+	return k, v
+}
+
+// First retrieves the first key value pair in the bucket.
+func (c *Cursor) First() ([]byte, []byte) {
+	if c.closed {
+		return nil, nil
+	}
+
+	k, v := c.cursor.First()
+	if len(k) == 0 && len(v) == 0 {
+		return nil, nil
+	}
+	return k, v
+}
+
+// Last retrieves the last key value pair in the bucket.
+func (c *Cursor) Last() ([]byte, []byte) {
+	if c.closed {
+		return nil, nil
+	}
+
+	k, v := c.cursor.Last()
+	if len(k) == 0 && len(v) == 0 {
+		return nil, nil
+	}
+	return k, v
+}
+
+// Next retrieves the next key in the bucket.
+func (c *Cursor) Next() (k []byte, v []byte) {
+	if c.closed {
+		return nil, nil
+	}
+	// get and unset previously seeked values if they exist
+	k, v, c.key, c.value = c.key, c.value, nil, nil
+	if len(k) > 0 && len(v) > 0 {
+		return
+	}
+
+	k, v = c.cursor.Next()
+	if len(k) == 0 && len(v) == 0 {
+		return nil, nil
+	}
+	return k, v
+}
+
+// Prev retrieves the previous key in the bucket.
+func (c *Cursor) Prev() (k []byte, v []byte) {
+	if c.closed {
+		return nil, nil
+	}
+	// get and unset previously seeked values if they exist
+	k, v, c.key, c.value = c.key, c.value, nil, nil
+	if len(k) > 0 && len(v) > 0 {
+		return
+	}
+
+	k, v = c.cursor.Prev()
+	if len(k) == 0 && len(v) == 0 {
+		return nil, nil
+	}
+	return k, v
+}
+
+// Err always returns nil as nothing can go wrong™ during iteration
```

Comment:
😆

Reply:
lol, that came from george [in influxdb](https://github.com/influxdata/influxdb/blame/48b8cb84f72b25a01e34587448418c5506f5f4e3/bolt/kv.go#L322) :smile: